internal/domain/power-status.go

```diff
@@ -18,16 +19,707 @@
+// pwrStateFilter:  Power state filter.
+// mgmtStateFilter: Management state filter.  
+// Return:          Passback object populated with model.PowerStatus object.
+
+func GetPowerStatus(xnames []string,
+                    pwrStateFilter pcsmodel.PowerStateFilter,
+                    mgmtStateFilter pcsmodel.ManagementStateFilter) (pb pcsmodel.Passback) {
+	//Grab all KVs from ETCD, cycle through them to make a map, then
+	//match the xnames in the passed-in array.  Grab pertinent data and create
+	//a pcsmodel.PowerStatus object 'pstatus', and return it.
+
+	statusObj,err := (*kvStore).GetAllPowerStatus()
+	if (err != nil) {
+		//TODO: we don't have an HTTP status code from a failed 
+		//GetAllPowerStatus() call; might need to pass that back in a 
+		//future mod.
+		return pcsmodel.BuildErrorPassback(http.StatusInternalServerError,err)
+	}
+
+	//Make a map of all returned components for xnames array matching.
+
+	compMap := make(map[string]*pcsmodel.PowerStatusComponent)
+
+	for ix,comp := range(statusObj.Status) {
+		compMap[comp.XName] = &statusObj.Status[ix]
+	}
+
+	//Build a return object, filtering by xname.
+
+	var rcomps pcsmodel.PowerStatus
+	var robj pcsmodel.Passback
+
+	psUndef := ((pwrStateFilter == pcsmodel.PowerStateFilter_Nil) ||
+			    (pwrStateFilter == pcsmodel.PowerStateFilter_Undefined))
+	msUndef := ((mgmtStateFilter == pcsmodel.ManagementStateFilter_Nil) ||
+			    (mgmtStateFilter == pcsmodel.ManagementStateFilter_undefined))
+
+	for _,name := range(xnames) {
+		stateMatch := true
+		mp,mapok := compMap[name]
+		if (!mapok) {
+			rcomps.Status = append(rcomps.Status,
+					pcsmodel.PowerStatusComponent{XName: name,
+						Error: "Component not found in component map."})
+			continue
+		}
+		//Filter by pwrstate and mgmtstate
+		if (!psUndef) {
+			if (pwrStateFilter.String() != mp.PowerState) {
+				stateMatch = false
+			}
+		}
+		if (!msUndef) {
+			if (mgmtStateFilter.String() != mp.ManagementState) {
+				stateMatch = false
+			}
+		}
+
+		if (stateMatch) {
+			var cmp pcsmodel.PowerStatusComponent
+			cmp.SupportedPowerTransitions = make([]string,len(mp.SupportedPowerTransitions))
+			cmp.XName = name
+			cmp.PowerState = mp.PowerState
+			cmp.ManagementState = mp.ManagementState
+			copy(cmp.SupportedPowerTransitions,mp.SupportedPowerTransitions)
+			rcomps.Status = append(rcomps.Status,cmp)
+		}
+	}
+
+	robj = pcsmodel.BuildSuccessPassback(200,rcomps)
+	return robj
+}
+
+/////////////////////////////////////////////////////////////////////////////
+//                I N T E R N A L   F U N C T I O N S
+/////////////////////////////////////////////////////////////////////////////
+
+// Update the cached HW state map using HSM.  This will add new components
+// to the map but not the actual HW states.
+
+func updateComponentMap() error {
+	fname := "updateComponentMap()"
+
+	//Get all components in HSM
+
+	compMap,err := (*hsmHandle).FillHSMData([]string{"all"})
```

Comment:
This routine is going to produce a lot of error messages in the log for no apparent reason. I would like to see a query to HSM for only the component types we care about. That way you don't have to filter later and drop large numbers of errors into the logs for components we don't want to cache.

Reply:
For example, here is the list of component types from Shandy
ncn-m001:~ # cray hsm state components list | jq '.[]|.[]|.Type' | sort | uniq -c
      2 "CabinetPDU"
      1 "CabinetPDUController"
     72 "CabinetPDUPowerConnector"
     32 "Chassis"
     32 "ChassisBMC"
    256 "ComputeModule"
    134 "HSNBoard"
   1036 "Node"
    523 "NodeBMC"
    523 "NodeEnclosure"
    134 "RouterBMC"
    256 "RouterModule"

Reply:
Here is the list from CAPMC and which you are missing.
"Node"
"ComputeModule" MISSING
"HSNBoard" MISSING
"RouterModule" MISSING
"Chassis"
"CabinetPDUOutlet" (I think this is the old name, below is the new one)
"CabinetPDUPowerConnector" MISSING

Reply:
This particular call is to HSM, so there shouldn't be any error messages.   I think the
missing types would come into play in your other comments.   I'll address them there.